packages/cmrjs/cmr-utils.js

```diff
@@ -274,50 +274,73 @@ function mapCNMTypeToCMRType(type) {
+async function generateFileUrl({
+  file,
+  distEndpoint,
+  cmrGranuleUrlType = 'distribution',
+  distributionBucketMap
+}) {
   if (cmrGranuleUrlType === 'distribution') {
-    const extension = urljoin(file.bucket, getS3KeyOfFile(file));
+    const bucketPath = _get(distributionBucketMap, file.bucket);
+    if (!bucketPath) {
+      throw new errors.MissingBucketMap(`No distribution bucket mapping exists for ${file.bucket}`);
+    }
+
+    const extension = urljoin(bucketPath, getS3KeyOfFile(file));
     return urljoin(distEndpoint, extension);
   }
 
   if (cmrGranuleUrlType === 's3') {
+    /* The check for file.filename is here
+       for legacy compliance reasons due to model simplification in
+       CUMULUS-1139 where filename was remapped to bucket and key*/
     if (file.filename) {
       return file.filename;
     }
-
     return buildS3Uri(file.bucket, file.key);
   }
 
-  return null;
+  return undefined;
 }
 
+
 /**
  * Construct online access url for a given file.
  *
  * @param {Object} params - input parameters
  * @param {Object} params.file - file object
  * @param {string} params.distEndpoint - distribution endpoint from config
  * @param {BucketsConfig} params.buckets -  stack BucketConfig instance
- * @returns {Object} online access url object
+ * @param {distributionBucketMap} params.distributionBucketMap - Object with bucket:tea-path mapping
+ *                                                               for all distribution bucketss
+ * @returns {(Object | undefined)} online access url object, undefined if no URL exists
  */
-function constructOnlineAccessUrl({
+async function constructOnlineAccessUrl({
   file,
   distEndpoint,
   buckets,
-  cmrGranuleUrlType = 'distribution'
+  cmrGranuleUrlType = 'distribution',
+  distributionBucketMap
 }) {
-  const bucketType = buckets.type(file.bucket);
-  const distributionApiBuckets = ['protected', 'public'];
-  const fileUrl = generateFileUrl(file, distEndpoint, cmrGranuleUrlType);
-  if (fileUrl && distributionApiBuckets.includes(bucketType)) {
-    return {
-      URL: fileUrl,
-      URLDescription: 'File to download', // used by ECHO10
-      Description: 'File to download', // used by UMMG
-      Type: mapCNMTypeToCMRType(file.type) // used by UMMG
-    };
+  try {
+    const bucketType = buckets.type(file.bucket);
+    const distributionApiBuckets = ['protected', 'public'];
+    if (distributionApiBuckets.includes(bucketType)) {
+      const fileUrl = await generateFileUrl({ file, distEndpoint, cmrGranuleUrlType, distributionBucketMap });
+      if (fileUrl) {
+        return {
+          URL: fileUrl,
+          URLDescription: 'File to download', // used by ECHO10
+          Description: 'File to download', // used by UMMG
+          Type: mapCNMTypeToCMRType(file.type) // used by UMMG
+        };
+      }
+    }
+    return undefined;
+  } catch (error) {
+    log.error(`Error constructing online access urls for ${JSON.stringify(file)}: ${error.message}`);
```

Comment:
Rather than logging here, can we move the log message to where the error is actually being handled?

Reply:
This is actually not useful, particularly given generateFileUrl throws a distinct error as it should.     Removed: dbe765bf1